tests/__init__.py

```diff
@@ -7,23 +7,29 @@
 )
 import os
 import platform
+import sys
 from typing import (
     TYPE_CHECKING,
     Final,
+    TypeVar,
 )
 
-import pandas as pd
-from pandas.util.version import Version
 import pytest
 
-from pandas._typing import T
+if sys.version_info < (3, 12):
+    import pandas as pd
+else:
+    with pytest.warns(DeprecationWarning, match="datetime.datetime.utcfromtimestamp"):
```

Comment:
I don't understand why this is needed. Why does pytest see the deprecation warning, but when we import pandas into a regular python interpreter, the warning isn't shown?

Reply:
pandas triggers a warning when importing on 3.12 - with `-Werror` - it errors (while pandas is not fully initialized).

Reply:
But why does pytest pick up the warning, but not from the python interpreter?

If I run `pytest` with `-W error`, I get the full stack trace showing that it is an import of `dateutil` from `timestamps.pyx` .  I just don't understand why a regular import of pandas isn't showing the warning.  What is pytest doing differently?

Reply:
It does, the warning is just not printed by default:

```sh
$ python -Wdefault -c "import pandas"
/var/home/twoertwein/.cache/pypoetry/virtualenvs/pandas-stubs-KrNUNgd3-py3.12/lib/python3.12/site-packages/dateutil/tz/tz.py:37: DeprecationWarning: datetime.datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.fromtimestamp(timestamp, datetime.UTC).
  EPOCH = datetime.datetime.utcfromtimestamp(0)
```


I will squash and force push to hopefully trigger the CI.

Reply:
just the random mac failure :)

Reply:
Looks like they fixed it but haven't released:  https://github.com/dateutil/dateutil/pull/1285

Reply:
Happy to wait until the next release then

Reply:
> just the random mac failure :)

I reran by hand and it worked.

Reply:
> Happy to wait until the next release then

They haven't released for over 2 years (last release was July, 2021), so I don't think we should wait.  If they ever do release, we can then revert that part of the change.